### ISSUE TITLE
False positive: redundant_clone

### ISSUE THREAD

**Issue by glittershark** [rust-lang/rust-clippy#10577]
### Summary

I looked and this didn't look like any of the existing `redundant_clone` false positives. Sometime between `nightly-2022-10-01` and `nightly-2023-03-18` (sorry for the wide range) `clippy::redundant_clone` started false-positive firing here:
https://github.com/readysettech/readyset/blob/82c3e22f64610cfdebed7d3f227ff477b25c5b10/readyset-server/src/controller/migrate/materialization/plan.rs#L277-L278

If that `clone()` call is removed, compilation fails due to `index_on` being used here:
https://github.com/readysettech/readyset/blob/82c3e22f64610cfdebed7d3f227ff477b25c5b10/readyset-server/src/controller/migrate/materialization/plan.rs#LL427C36-L427C44

### Lint Name

clippy::redundant_clone

### Reproducer

See links

### Version

```text
❯ rustc -Vv
rustc 1.70.0-nightly (4a04d086c 2023-03-18)
binary: rustc
commit-hash: 4a04d086cac54a41517d5657b59d5fe2caca2d71
commit-date: 2023-03-18
host: x86_64-unknown-linux-gnu
release: 1.70.0-nightly
LLVM version: 16.0.0
```

**Comment by ilyagr** [comment]
I don't think this triggered as recently as `rustc 1.69.0-nightly (13471d3b2 2023-03-02)`. At least, that's true for my similar issue at https://github.com/martinvonz/jj/pull/1465.

**Comment by andrey-yantsen** [comment]
To help narrow down the range — the issue started sometime between 2023-03-16 and 2023-03-18. The [last successful build](https://github.com/andrey-yantsen/plex-api.rs/actions/runs/4440989409/jobs/7795487573) for me was on 16 March 2023, and [the first failed](https://github.com/andrey-yantsen/plex-api.rs/actions/runs/4456020640/jobs/7826193713) was on 18 March.

**Comment by zeenix** [comment]
FWIW I bumped into this too with [this code](https://gitlab.freedesktop.org/dbus/zbus/-/blob/main/zbus_macros/src/error.rs#L239) in zbus.

```
error: redundant clone
   --> zbus_macros/src/error.rs:239:26
    |
239 |                 in_fields.clone()
    |                          ^^^^^^^^ help: remove this
    |
note: cloned value is neither consumed nor mutated
   --> zbus_macros/src/error.rs:239:17
    |
239 |                 in_fields.clone()
    |                 ^^^^^^^^^^^^^^^^^
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_clone
    = note: `-D clippy::redundant-clone` implied by `-D warnings`
```

**Comment by ryoqun** [comment]
hey, i created minimal test case:

```rust
use std::sync::Arc;

#[derive(Clone)]
struct Account {
    data: Arc<usize>,
}

fn take<T>(_a: &T) {
}

fn main() {
    let account = Account {data: Arc::new(30)};
    let data = account.data.clone();
    take(&account);
    take(&data);
}

```

```
$ cargo +nightly-2023-03-16-x86_64-unknown-linux-gnu clippy --fix --allow-dirty -- --deny clippy::redundant_clone
    Checking clippy-redundant-clone v0.1.0 (/home/sol/work/clippy-redundant-clone)
    Finished dev [unoptimized + debuginfo] target(s) in 0.26s

$ cargo +nightly-2023-03-17-x86_64-unknown-linux-gnu clippy --fix --allow-dirty -- --deny clippy::redundant_clone
    Checking clippy-redundant-clone v0.1.0 (/home/sol/work/clippy-redundant-clone)
warning: failed to automatically apply fixes suggested by rustc to crate `clippy_redundant_clone`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0382]: borrow of partially moved value: `account`
  --> src/main.rs:14:10
   |
13 |     let data = account.data;
   |                ------------ value partially moved here
14 |     take(&account);
   |          ^^^^^^^^ value borrowed here after partial move
   |
   = note: partial move occurs because `account.data` has type `std::sync::Arc<usize>`, which does not implement the `Copy` trait

error: aborting due to previous error

For more information about this error, try `rustc --explain E0382`.
Original diagnostics will follow.

warning: redundant clone
  --> src/main.rs:13:28
   |
13 |     let data = account.data.clone();
   |                            ^^^^^^^^ help: remove this
   |
note: cloned value is neither consumed nor mutated
  --> src/main.rs:13:16
   |
13 |     let data = account.data.clone();
   |                ^^^^^^^^^^^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_clone
   = note: requested on the command line with `-D clippy::redundant-clone`

warning: `clippy-redundant-clone` (bin "clippy-redundant-clone") generated 1 warning (run `cargo clippy --fix --bin "clippy-redundant-clone"` to apply 1 suggestion)
warning: `clippy-redundant-clone` (bin "clippy-redundant-clone" test) generated 1 warning (1 duplicate)
    Finished dev [unoptimized + debuginfo] target(s) in 0.25s
```

**Comment by ryoqun** [comment]
also, i further narrowed down the regression window (thank to the prior work: https://github.com/rust-lang/rust-clippy/issues/10577#issuecomment-1494685235):

so, it looks like the regression window is quite narrow: https://github.com/rust-lang/rust/compare/ab65486...511364e

this is a hunch but i suspect https://github.com/rust-lang/rust/pull/108944 is the culprit, flipping this line: https://github.com/rust-lang/rust-clippy/blob/c4909746add9b34e3f10508ca37b87f5aa96c0b8/clippy_lints/src/redundant_clone.rs#L381

**Comment by printfn** [comment]
It looks like this bug is now in stable 1.70, I'm getting errors in https://github.com/printfn/fend.

<details>

```
error: redundant clone
   --> core/src/num/biguint.rs:199:34
    |
199 |         let mut high_guess = self.clone();
    |                                  ^^^^^^^^ help: remove this
    |
note: cloned value is neither consumed nor mutated
   --> core/src/num/biguint.rs:199:30
    |
199 |         let mut high_guess = self.clone();
    |                              ^^^^^^^^^^^^
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_clone
note: the lint level is defined here
   --> core/src/lib.rs:2:9
    |
2   | #![deny(clippy::all)]
    |         ^^^^^^^^^^^
    = note: `#[deny(clippy::redundant_clone)]` implied by `#[deny(clippy::all)]`

```

</details>

**Comment by JarredAllen** [comment]
I'm also seeing this error, with an even simpler minimal test case than was produced above (no need for `Arc`s): [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d09804170d27125a2655170078a9239d)

**Comment by ryoqun** [comment]
yeah, fyi, I confirmed that the root cause is rustc's internal changes (ref: https://github.com/rust-lang/rust/pull/108944#issuecomment-1564451009). hehe, as i guessed in the post, this is getting more attention because now 1.70 is shipped. ;)

according to https://github.com/rust-lang/rust/pull/108944#issuecomment-1564493704, it seems clippy needs to be fixed for its misuse somehow...

fyi, i only bisected as a normal rust community member, currently not actively trying fix the clippy.

**Comment by foresterre** [comment]
Another instance in indicatif: https://github.com/rust-lang/rust/issues/112227

**Comment by AaronKutch** [comment]
Another set of instances is in my crate `awint` as of commit https://github.com/AaronKutch/awint/commit/71b35e8f034126d416d435fd718a2a6b95e94c4a (run `cargo clippy --no-default-features`), the clone isn't redundant because the original var needs to be used for a special trait method call

**Comment by joshtriplett** [comment]
Since this seems to be causing a number of false positives, could we temporarily mark it as allow-by-default until it can be fixed?
